actions/gha-workflow-validator/src/utils.ts

```diff
@@ -0,0 +1,165 @@
+  for (const line of lineChanges) {
+
+    if (line.startsWith('@@')) {
+      // @@ denotes a git hunk header
+      // https://mattscodecave.com/posts/howto-reading-git-diffs-and-staging-hunks.html
+      // example line: @@ -16,10 +16,10 @@ jobs:
+      //   - "-16,10": "-" denotes source file, "16,10" means the hunk starts at line 16 and output contains 10 lines
+      //   - "+16,10": "+" denotes destination file, "16,10" means the same as above
+      const [ , , dest, ] = line.split(' ');
+
+      if (!dest.startsWith("+")) {
+        throw new Error("Invalid git hunk format");
+      }
+
+      const [ destLine, ] = dest.substring(1).split(',');
+      currentLineInFile = parseInt(destLine, 10);
+      continue;
+    } else if (line.startsWith('+')) {
+      const currentLine = line.substring(1);
+      const actionReference = extractActionReference(currentLine);
+      additions.push({ content: currentLine, lineNumber: currentLineInFile, actionReference });
+    } else if (line.startsWith('-')) {
+      // ignore deletions
+      continue;
+    }
+    currentLineInFile++;
+  }
+
+  return additions;
+}
+
+function extractActionReference(line: string): ActionReference | undefined {
+  // example line:
+  //       - uses: actions/checkout@9bb56186c3b09b4f86b1c65136769dd318469633 # v4.1.2
+  const trimSubString = "uses:"
+  const usesIndex = line.indexOf(trimSubString);
+
+  if (usesIndex === -1) {
+    // Not an action reference
+    return;
+  }
+
+  // trim past the "uses:" substring to get "<owner>/<repo><optional path>@<ref> # <optional comment>""
+  const trimmedLine = line.substring(line.indexOf(trimSubString) + trimSubString.length).trim();
+
+  if (trimmedLine.startsWith("./")) {
+    // Local action reference - do not validate these yet (TODO)
+    return;
+  }
+
+  const [ actionIdentifier, ...comment ] = trimmedLine.split("#");
+  const [ identifier, gitRef ] = actionIdentifier.trim().split("@");
+  const [ owner, repo, ...path] = identifier.split("/");
+  const repoPath = ((path.length) > 0 ? "/" : "") + path.join("/");
+
+  return { owner, repo, repoPath, ref: gitRef, comment: comment.join().trim(), line, };
+}
+
+export function annotatePR(validationResults: ValidationResult[]) {
+  for (const fileResults of validationResults) {
+    for (const lineResults of fileResults.lineValidations) {
+      const message = lineResults.validationErrors.map(error => error.message).join("\n");
+      core.error(message, {
+        file: fileResults.filename,
+        startLine: lineResults.line.lineNumber,
+      });
+    }
+  }
+}
+
+type TableRow = Parameters<typeof core.summary.addTable>[0][0]
+type TableCell = TableRow[0]
+
+export async function setSummary(validationResults: ValidationResult[], fileUrlPrefix: string) {
+  const headerRows: TableRow[] = [
+    // Header Row
+    [
+      { data: "Filename", header: true },
+      { data: "Line Number", header: true },
+      { data: "Violations", header: true },
+    ],
+  ];
```

Comment:
```suggestion
  const headerRow: TableRow = 
     [
      { data: "Filename", header: true },
      { data: "Line Number", header: true },
      { data: "Violations", header: true },
  ];
```

Reply:
I'll remove the comment, idk what else is happening here?